servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/BlockingProcessorBuffer.java

```diff
@@ -0,0 +1,90 @@
+
+    /**
+     * Adds an item to this buffer.
+     *
+     * @param item to add.
+     * @throws InterruptedException If the add was interrupted.
+     */
+    void add(@Nullable T item) throws InterruptedException;
+
+    /**
+     * Terminates this buffer, such that no further modifications of this buffer are allowed. Subsequent
+     * {@link BufferConsumer consumptions} must first consume all previously {@link #add(Object) added} items and then
+     * {@link BufferConsumer#consumeTerminal()}  consume termination}.
+     * @throws InterruptedException If termination was interrupted.
+     */
+    void terminate() throws InterruptedException;
+
+    /**
+     * Terminates this buffer, such that no further modifications of this buffer are allowed. Subsequent
+     * {@link BufferConsumer consumptions} must first consume all previously {@link #add(Object) added} items and then
+     * {@link BufferConsumer#consumeTerminal()}  consume termination}.
+     *
+     * @param cause {@link Throwable} as a cause for termination.
+     * @throws InterruptedException If termination was interrupted.
+     */
+    void terminate(Throwable cause) throws InterruptedException;
+
+    /**
+     * Consumes the next item stored in this buffer. If there are no items stored in the buffer and the buffer has
+     * terminated {@link #terminate() successfully} or with an {@link #terminate(Throwable) error} then consume that
+     * {@link BufferConsumer#consumeTerminal() successful} or {@link BufferConsumer#consumeTerminal(Throwable) failed}
+     * termination.
+     * <p>
+     * This method will block till an item or a terminal event is available in the buffer.
+     *
+     * @param consumer {@link BufferConsumer} to consume the next item or termination in this buffer
+     * @return {@code true} if any method was called on the passed {@link BufferConsumer}.
+     * @throws InterruptedException If the thread was interrupted while waiting for an item or terminal event.
+     */
+    boolean consume(BufferConsumer<T> consumer) throws InterruptedException;
+
+    /**
+     * Consumes the next item stored in this buffer. If there are no items stored in the buffer and the buffer has
+     * terminated {@link #terminate() successfully} or with an {@link #terminate(Throwable) error} then consume that
+     * {@link BufferConsumer#consumeTerminal() successful} or {@link BufferConsumer#consumeTerminal(Throwable) failed}
+     * termination.
+     * <p>
+     * This method will block till an item or a terminal event is available in the buffer or the passed {@code waitFor}
+     * duration has elapsed.
+     *
+     * @param consumer {@link BufferConsumer} to consume the next item or termination in this buffer
+     * @param waitFor Duration to wait for an item or termination to be available.
+     * @param waitForUnit {@link TimeUnit} for {@code waitFor}.
+     * @return {@code true} if any method was called on the passed {@link BufferConsumer}.
```

Comment:
is this really true ? Like will it also return true if `consume` was called ?

Reply:
Yes that is correct. Return value is indicating whether an item or terminal was consumed.